cinder_plugin/tests/test_volume.py

```diff
@@ -339,4 +340,317 @@ def test_detach(self, wait_until_status_m, cinder_m, nova_m):
+    def test_snapshot_apply(self, cinder_m):
+        cinder_instance = cinder_m.return_value
+        volume_ctx, volume_id = self._simple_volume_ctx()
+
+        # Snapshot
+        cinder_instance.backups.list = mock.Mock(return_value=[])
+        cinder_instance.volume_snapshots.list = mock.Mock(return_value=[])
+
+        volume.snapshot_apply(ctx=volume_ctx, snapshot_name="snapshot_name",
+                              snapshot_incremental=True)
+
+        cinder_instance.backups.list.assert_not_called()
+        cinder_instance.volume_snapshots.list.assert_not_called()
+
+        # No such backup
+        cinder_instance.backups.list = mock.Mock(return_value=[])
+        cinder_instance.volume_snapshots.list = mock.Mock(return_value=[])
+
+        with self.assertRaises(cfy_exc.NonRecoverableError):
+            volume.snapshot_apply(ctx=volume_ctx, snapshot_name="backup_name",
+                                  snapshot_incremental=False)
+
+        cinder_instance.backups.list.assert_called_once_with(
+            search_opts={
+                'name': 'vol-1234-5678-backup_name',
+                'volume_id': '1234-5678'})
+        cinder_instance.volume_snapshots.list.assert_not_called()
+
+        # backup exist
+        backup_mock = mock.Mock()
+        backup_mock.name = 'vol-1234-5678-backup_name'
+        backup_mock.id = 'backup_id'
+        cinder_instance.restores.restore = mock.Mock()
+        cinder_instance.backups.list = mock.Mock(return_value=[backup_mock])
+
+        volume.snapshot_apply(ctx=volume_ctx, snapshot_name="backup_name",
+                              snapshot_incremental=False)
+
+        cinder_instance.backups.list.assert_called_once_with(
+            search_opts={
+                'name': 'vol-1234-5678-backup_name',
+                'volume_id': '1234-5678'})
+        cinder_instance.restores.restore.assert_called_once_with(
+            'backup_id', '1234-5678')
+        cinder_instance.volume_snapshots.list.assert_not_called()
+
+    @mock.patch('openstack_plugin_common.CinderClientWithSugar')
+    @mock.patch('time.sleep', mock.Mock())
+    def test_list_volumes(self, cinder_m):
+        cinder_instance = cinder_m.return_value
+        volume_ctx, volume_id = self._simple_volume_ctx()
+
+        cinder_instance.volumes.list = mock.Mock(return_value=[])
+
+        volume.list_volumes(ctx=volume_ctx, args={"abc": "def"})
+
+        cinder_instance.volumes.list.assert_called_once_with(abc="def")
+        self.assertEqual(
+            {'external_id': '1234-5678', 'volume_list': []},
+            volume_ctx.instance.runtime_properties
+        )
+
+    @mock.patch('openstack_plugin_common.CinderClientWithSugar')
+    def test_creation_validation(self, cinder_m):
+        volume_ctx, volume_id = self._simple_volume_ctx()
+
+        volume.creation_validation(ctx=volume_ctx)
+
+    @mock.patch('openstack_plugin_common.CinderClientWithSugar')
+    @mock.patch('time.sleep', mock.Mock())
```

Comment:
@0lvin same here for time

Reply:
@mabuaisha code removes volume and for case "fast" remove recheck that have deleted snapshot after small sleep. And only after that raise exception.